### PR TITLE
[backport] Refactor array testing decorators

### PR DESCRIPTION
**Pull request by emcastillo**
Backport of #2818

### CODE REVIEW

**Comment by emcastillo** [comment]
[automatic post] Jenkins, test this please.

**Comment by emcastillo** [comment]
Seems like #2394 should be backported to v6 first 🤔,
or do we just not merge this in v6.

**Comment by niboshi** [comment]
I think #2498 should be backported instead.

Current v6 has a problem that the decorators does not treat `SkipTest` correctly, so if cupy raises it, the test will be skipped even if numpy behaves differently. That shoud be fixed (by backporting #2498).

However, #2498 alone does not fix all decorators. As `numpy_cupy_array_list_equal()` is not using the common implementation, #2498 won't fix the decorator. Backporting #2498 and this PR together fixes all the decorators.

**Comment by emcastillo** [comment]
Backported #2892

File: cupy/testing/helper.py [hunk 1]
```diff
@@ -40,6 +39,91 @@ def _call_func(self, impl, args, kw):
...
+    cupy_axis_error = cupy.core._errors._AxisError
+    if isinstance(cupy_error, cupy_axis_error) and numpy_axis_error is None:
+        if not isinstance(numpy_error, (ValueError, IndexError)):
+            return False
+        errors = list(set(errors) - set([IndexError, ValueError]))
+
+    return all([isinstance(cupy_error, err) == isinstance(numpy_error, err)
+                for err in errors])
+
+
```
Comment (niboshi) [inline cupy/testing/helper.py]: Please remove them.

**Comment by emcastillo** [comment]
PTAL!

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
LGTM

**Pull request merged**